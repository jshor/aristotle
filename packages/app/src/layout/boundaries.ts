function isInNeighborhood (a: Point, b: Point, radius: number): boolean {
  const distX = Math.pow(a.x - b.x, 2)
  const distY = Math.pow(a.y - b.y, 2)

  return Math.sqrt(distX + distY) <= radius
}

/**
 * Returns true if line a intersects with line B.
 *
 * @param a1 - first point of a
 * @param a2 - second point of a
 * @param b1 - first point of b
 * @param b2 - second point of b
 */
function hasLinearIntersection (a1: Point, a2: Point, b1: Point, b2: Point): boolean {
  const ccw = (a: Point, b: Point, c: Point) => {
    return (c.y-a.y) * (b.x-a.x) > (b.y-a.y) * (c.x-a.x)
  }

  return ccw(a1, b1, b2) !== ccw(a2, b1, b2) && ccw(a1, a2, b1) !== ccw(a1, a2, b2)
}

function isLineIntersectingRectangle (a1: Point, a2: Point, boundingBox: BoundingBox): boolean {
  const b1 = { x: boundingBox.left, y: boundingBox.top }
  const b2 = { x: boundingBox.right, y: boundingBox.top }
  const b3 = { x: boundingBox.right, y: boundingBox.bottom }
  const b4 = { x: boundingBox.left, y: boundingBox.bottom }

  return hasLinearIntersection(a1, a2, b1, b2)
    || hasLinearIntersection(a1, a2, b2, b3)
    || hasLinearIntersection(a1, a2, b3, b4)
    || hasLinearIntersection(a1, a2, b4, b1)
}

function scaleBoundingBox (box: BoundingBox, scale: number): BoundingBox {
  return {
    left: box.left * scale,
    top: box.top * scale,
    bottom: box.bottom * scale,
    right: box.right * scale
  }
}

function isTwoDimensional (box: BoundingBox) {
  return box.left !== box.right || box.top !== box.bottom
}

/**
 * Returns a list of one infinitely-horizontal and one infinitely-vertical boundaries.
 *
 * @param point - the point to compute boundaries for
 * @returns list of bounding boxes
 */
function getLinearBoundaries (point: Point): BoundingBox[] {
  return [
    {
      left: point.x,
      right: point.x,
      top: -Infinity,
      bottom: Infinity
    },
    {
      left: -Infinity,
      right: Infinity,
      top: point.y,
      bottom: point.y
    }
  ]
}

/**
 * Returns a point as a boundary.
 *
 * @param point
 * @returns point boundary
 */
function getPointBoundary (point: Point): BoundingBox {
  return {
    left: point.x,
    top: point.y,
    right: point.x,
    bottom: point.y
  }
}

/**
 * Returns true if a and b intersect.
 *
 * @param a
 * @param b
 * @returns boolean result of intersection
 */
function hasIntersection (a: BoundingBox, b: BoundingBox): boolean {
  if (a.left >= b.right || b.left >= a.right) {
    // one rect is on left side of other
    return false
  }

  if (a.top >= b.bottom || b.top >= a.bottom) {
    // one rect is above other
    return false
  }

  return true
}

/**
 * Computes the bounding box of the given item.
 *
 * @param item
 * @returns computed bounding box
 */
function getItemBoundingBox (item: Item): BoundingBox {
  if (item.rotation % 2 === 0) {
    // item is right side up or upside down
    return {
      left: item.position.x,
      top: item.position.y,
      bottom: item.height + item.position.y,
      right: item.width + item.position.x
    }
  }

  // item is rotated at a 90 degree angle (CW or CCW)
  const midX = item.position.x + (item.width / 2)
  const midY = item.position.y + (item.height / 2)

  return {
    left: midX - (item.height / 2),
    top: midY - (item.width / 2),
    right: midX + (item.height / 2),
    bottom: midY + (item.width / 2)
  }
}

/**
 * Computes the group bounding box for the given list of bounding boxes.
 *
 * @param boundingBoxes - list of bounding boxes contained
 * @returns computed bounding box
 */
function getGroupBoundingBox (boundingBoxes: BoundingBox[]): BoundingBox {
  return boundingBoxes.reduce((rect, boundingBox) => ({
    left: Math.min(rect.left, boundingBox.left),
    top: Math.min(rect.top, boundingBox.top),
    right: Math.max(rect.right, boundingBox.right),
    bottom: Math.max(rect.bottom, boundingBox.bottom)
  }), {
    left: Infinity,
    top: Infinity,
    right: 0,
    bottom: 0
  })
}

export default {
  isInNeighborhood,
  isLineIntersectingRectangle,
  scaleBoundingBox,
  isTwoDimensional,
  getLinearBoundaries,
  getPointBoundary,
  hasIntersection,
  getItemBoundingBox,
  getGroupBoundingBox
}
