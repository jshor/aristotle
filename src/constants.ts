/** Minimum zoom percentage. */
export const MIN_ZOOM = 0.1

/** Maximum zoom percentage. */
export const MAX_ZOOM = 2

/** The incremental percentage amount per zoom operation. */
export const ZOOM_STEP = 0.1

/** The frictional coefficient for animated panning movements. */
export const PANNING_FRICTION = 4

/** The speed scalar for animated panning movements using momentum. */
export const PANNING_SPEED = 15

/** The amount of padding, in pixels, for an exported image. */
export const IMAGE_PADDING = 20

/** The amount of momentum to build with each key press while moving an object using arrow keys. */
export const ARROW_KEY_MOMENTUM_MULTIPLIER = 1.05

/** The amount of time, in milliseconds, that the user must be touching something before emitting a `touchhold` event. */
export const TOUCH_HOLD_TIMEOUT = 100

/**
 * The minimum CSS z-index value to apply to all items.
 * This high number is to ensure that all items visually overlap all wires.
 */
export const ITEM_BASE_Z_INDEX = 1000

/** The easing function for each animation frame to ease out after panning. */
export const PANNING_EASING_FUNCTION = (x: number) => 1 - Math.pow(1 - x, 3)
