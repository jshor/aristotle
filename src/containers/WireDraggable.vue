<template>
  <svg
    class="wire"
    :class="{
      'wire--selected': isSelected,
      'wire--preview': isPreview,
      'wire--flash': flash
    }"
    :width="geometry.width"
    :height="geometry.height"
    aria-roledescription="connection"
  >
    <path
      class="wire__outline"
      fill="none"
      :transform="`translate(${Math.abs(geometry.minX)}, ${Math.abs(geometry.minY)})`"
      :d="geometry.path"
    />
    <path
      class="wire__display"
      :class="{
        'wire__display--on': sourceValue === LogicValue.TRUE,
        'wire__display--off': sourceValue === LogicValue.FALSE
      }"
      fill="none"
      :transform="`translate(${Math.abs(geometry.minX)}, ${Math.abs(geometry.minY)})`"
      :d="geometry.path"
    />
    <path
      class="wire__clickable"
      fill="none"
      ref="wireRef"
      :transform="`translate(${Math.abs(geometry.minX)}, ${Math.abs(geometry.minY)})`"
      :d="geometry.path"
      @mousedown.left="onLeftMouseDown"
      @mouseup="onMouseUp"
      @touchstart="onTouchStart"
      @touchmove="onTouchMove"
      @touchend="onTouchEnd"
      @touchcancel="dragEnd"
    />
  </svg>
</template>

<script lang="ts">
import LogicValue from '@/types/enums/LogicValue'
import WireGeometry from '@/types/types/WireGeometry'
import { computed, defineComponent, PropType, ref } from 'vue'
import { useDraggable } from '@/composables/useDraggable'
import Point from '@/types/interfaces/Point'

export default defineComponent({
  name: 'Wire',
  props: {
    /** The wire Bezier curve geometry. */
    geometry: {
      type: Object as PropType<WireGeometry>,
      required: true
    },
    isAnimated: {
      type: Boolean,
      default: false
    },
    sourceValue: {
      type: Number as PropType<LogicValue>,
      required: true
    },
    label: {
      type: String,
      default: ''
    },
    isPreview: {
      type: Boolean,
      default: false
    },
    isSelected: {
      type: Boolean,
      default: false
    },
    /** Whether or not the item should show a flash once to the user. */
    flash: {
      type: Boolean,
      default: false
    }
  },
  emits: {
    add: (point: Point, offset: Point) => true,
    move: (point: Point, offset: Point) => true,
    added: () => true,
    select: (deselectAll?: boolean) => true,
    deselect: () => true
  },
  setup (props, { emit }) {
    const wireRef = ref<SVGPathElement>()
    const isTouchDragging = ref(false)

    /**
     * Touch termination event handler.
     */
    function onTouched ($event: TouchEvent, held: boolean) {
      if (held) {
        isTouchDragging.value = true

        return emit('add', {
          x: $event.touches[0].clientX,
          y: $event.touches[0].clientY
        }, { x: 0, y: 0 })
      }

      props.isSelected
        ? emit('deselect')
        : emit('select', true)
    }

    function onDragStart (position: Point, offset: Point) {
      if (!isTouchDragging.value) {
        emit('add', position, offset)
      }
    }

    function onDragEnd () {
      isTouchDragging.value = false
      emit('added')
    }

    /**
     * Handles the left mouse down event.
     * This will select the item if it is not already selected.
     * The draggable `mousedown` event will continue to be propagated.
     */
    function onLeftMouseDown ($event: MouseEvent) {
      onMouseDown($event)

      if (!props.isSelected) {
        emit('select', !$event.ctrlKey)
      }
    }

    const allowTouchDrag = computed(() => isTouchDragging.value)
    const {
      onMouseDown,
      onMouseUp,
      onTouchStart,
      onTouchMove,
      onTouchEnd,
      dragEnd,
    } = useDraggable({
      allowTouchDrag,
      longTouchTimeout: 500,
      onTouched,
      onDragStart,
      onDrag: (p: Point, o: Point) => emit('move', p, o),
      onDragEnd: () => onDragEnd()
    })

    return {
      wireRef,
      LogicValue,
      onLeftMouseDown,
      onMouseUp,
      onTouchStart,
      onTouchMove,
      onTouchEnd,
      dragEnd
    }
  }
})
</script>

<style lang="scss">
.wire {
  @include circuit-element;

  pointer-events: none;
  touch-action: none;
  position: absolute;

  &--preview {
    opacity: 0.5;
  }

  &__clickable {
    animation: none;
    stroke-width: 16;
    pointer-events: visibleStroke;
    cursor: pointer;
  }

  &__outline {
    stroke-width: 8;
    stroke-linecap: square;
    stroke: var(--color-secondary);
  }

  &__animation {
    stroke-width: 3;
    stroke-dasharray: 5;
    animation: animate1 500s infinite linear;
    stroke: red;
  }

  &__display {
    stroke-width: 6;
    stroke: var(--color-hi-z);
    transition: 0.25s stroke; // TODO

    &--on {
      stroke: var(--color-on);
      stroke: -webkit-focus-ring-color !important;
    }

    &--off {
      stroke: var(--color-off);
      stroke: -webkit-focus-ring-color !important;
    }
  }
}

@keyframes animate1 {
  from {
    stroke-dashoffset: 0;
  }
  to {
    stroke-dashoffset: -10000;
  }
}

@keyframes animate2 {
  from {
    stroke-dashoffset: 0;
  }
  to {
    stroke-dashoffset: 0;
  }
}

@keyframes flash { // TODO: make this a mixin
  0% {
    filter: inherit;
  }
  50% {
    filter: drop-shadow(0px 0px 20px var(--color-primary));
  }
  100% {
    filter: inherit;
  }
}
</style>
