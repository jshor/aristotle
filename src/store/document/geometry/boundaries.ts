import SnapMode from "@/types/enums/SnapMode"
import Point from "@/types/interfaces/Point"
import BoundingBox from "@/types/types/BoundingBox"

/**
 * Rounds x to the nearest n.
 */
function roundToNth (x: number, n: number) {
  return Math.round(x / n) * n
}

/**
 * Returns true if a is within the radial distance of b.
 */
function isInNeighborhood (a: Point, b: Point, radius: number): boolean {
  return Math.hypot(a.x - b.x, a.y - b.y) <= radius
}

/**
 * Returns the delta of rounding the top-left point of the given bounding box to the nearest distance.
 * This will achieve the effect of "snapping to grid."
 *
 * @param a - the bounding box of the element moving
 * @param distance - the grid size
 */
function getGridSnapPosition (a: BoundingBox, distance: number): Point {
  return {
    x: roundToNth(a.left, distance) - a.left,
    y: roundToNth(a.top, distance) - a.top
  }
}

/**
 * Returns the offset Cartesian distance between bounding box `a` and any other bounding boxes within the radial `distance`.
 * If no bounding boxes are within 'd' distance, it returns an offset of (0, 0).
 *
 * @param boundingBoxes
 * @param a
 * @param d
 */
function getRadialSnapOffset (boundingBoxes: BoundingBox[], a: BoundingBox, distance: number): Point {
  for (let i = 0; i < boundingBoxes.length; i++) {
    const b = boundingBoxes[i]

    if (isInNeighborhood({ x: b.left, y: b.top }, { x: a.left, y: a.top }, distance)) {
      return {
        x: b.left - a.left,
        y: b.top - a.top
      }
    }
  }

  return { x: 0, y: 0 }
}

/**
 * Returns the offset Cartesian distance between bounding box `a` and any other bounding boxes within the linear `distance`.
 * If no bounding boxes are within 'd' distance, it returns an offset of (0, 0).
 *
 * @param boundingBoxes
 * @param a
 * @param d
 */
function getOuterSnapOffset (boundingBoxes: BoundingBox[], a: BoundingBox, distance: number): Point {
  const offset: Point = { x: 0, y: 0 }

  for (let i = 0; i < boundingBoxes.length; i++) {
    const b = boundingBoxes[i]
    const ts = Math.abs(b.top - a.bottom) <= distance
    const bs = Math.abs(b.bottom - a.top) <= distance
    const ls = Math.abs(b.left - a.right) <= distance
    const rs = Math.abs(b.right - a.left) <= distance

    if (
      (a.left <= b.right && a.left >= b.left) ||
      (a.right >= b.left && a.right <= b.right) ||
      (a.left <= b.left && a.right >= b.right)
    ) {
      // a is within the y-axis boundaries
      if (ts) offset.y = b.top - a.bottom
      if (bs) offset.y = b.bottom - a.top
    }

    if (
      (a.top <= b.bottom && a.top >= b.top) ||
      (a.bottom >= b.top && a.bottom <= b.bottom) ||
      (a.top <= b.top && a.bottom >= b.bottom)
    ) {
      // a is within the x-axis boundaries
      if (ls) offset.x = b.left - a.right
      if (rs) offset.x = b.right - a.left
    }
  }

  return offset
}

function getSnapOffset (boundingBoxes: BoundingBox[], boundingBox: BoundingBox, snapMode: SnapMode, distance: number) {
  switch (snapMode) {
    case SnapMode.Radial:
      return getRadialSnapOffset(boundingBoxes, boundingBox, distance)
    case SnapMode.Grid:
      return getGridSnapPosition(boundingBox, distance)
    case SnapMode.Outer:
    default:
      return getOuterSnapOffset(boundingBoxes, boundingBox, distance)
  }
}

/**
 * Returns true if line a intersects with line B.
 *
 * @param a1 - first point of a
 * @param a2 - second point of a
 * @param b1 - first point of b
 * @param b2 - second point of b
 */
function hasLinearIntersection (a1: Point, a2: Point, b1: Point, b2: Point): boolean {
  const ccw = (a: Point, b: Point, c: Point) => {
    return (c.y-a.y) * (b.x-a.x) > (b.y-a.y) * (c.x-a.x)
  }

  return ccw(a1, b1, b2) !== ccw(a2, b1, b2) && ccw(a1, a2, b1) !== ccw(a1, a2, b2)
}

/**
 * Returns true if the lines between the given connection points intersect with the given bounding box.
 */
function areLinesIntersectingRectangle (points: Point[], boundingBox: BoundingBox): boolean {
  let previousPoint = points[0]

  for (let i = 1; i < points.length; i++) {
    if (isLineIntersectingRectangle(previousPoint, points[i], boundingBox)) {
      return true
    }

    previousPoint = points[i]
  }

  return false
}

/**
 * Returns true if the overlaps with the given bounding box.
 *
 * @param a1 - first point of the line
 * @param a2 - second point of the line
 * @param boundingBox - the rectange
 */
function isLineIntersectingRectangle (a1: Point, a2: Point, boundingBox: BoundingBox): boolean {
  const b1 = { x: boundingBox.left, y: boundingBox.top }
  const b2 = { x: boundingBox.right, y: boundingBox.top }
  const b3 = { x: boundingBox.right, y: boundingBox.bottom }
  const b4 = { x: boundingBox.left, y: boundingBox.bottom }

  return hasLinearIntersection(a1, a2, b1, b2)
    || hasLinearIntersection(a1, a2, b2, b3)
    || hasLinearIntersection(a1, a2, b3, b4)
    || hasLinearIntersection(a1, a2, b4, b1)
}

/**
 * Returns true if a and b intersect.
 *
 * @param a
 * @param b
 * @returns boolean result of intersection
 */
function hasIntersection (a: BoundingBox, b: BoundingBox): boolean {
  if (a.left >= b.right || b.left >= a.right) {
    // one rect is on left side of other
    return false
  }

  if (a.top >= b.bottom || b.top >= a.bottom) {
    // one rect is above other
    return false
  }

  return true
}

/**
 * Scales the bounding box by the given scalar.
 *
 * @param box
 * @param scale
 */
function scaleBoundingBox (box: BoundingBox, scale: number): BoundingBox {
  return {
    left: box.left * scale,
    top: box.top * scale,
    bottom: box.bottom * scale,
    right: box.right * scale
  }
}

/**
 * Returns true if the given bounding box is one- or two-dimensional.
 */
function isOneOrTwoDimensional (box: BoundingBox) {
  return box.left !== box.right || box.top !== box.bottom
}

/**
 * Returns a list of one infinitely-horizontal and one infinitely-vertical boundaries.
 *
 * @param point - the point to compute boundaries for
 * @returns list of bounding boxes
 */
function getLinearBoundaries (point: Point): BoundingBox[] {
  return [
    {
      left: point.x,
      right: point.x,
      top: -Infinity,
      bottom: Infinity
    },
    {
      left: -Infinity,
      right: Infinity,
      top: point.y,
      bottom: point.y
    }
  ]
}

/**
 * Returns a point as a boundary.
 *
 * @param point
 * @returns point boundary
 */
function getPointBoundary (point: Point): BoundingBox {
  return {
    left: point.x,
    top: point.y,
    right: point.x,
    bottom: point.y
  }
}

/**
 * Computes the bounding box of the given item.
 *
 * @param item
 * @returns computed bounding box
 */
function getBoundingBox (position: Point, rotation: number, width: number, height: number): BoundingBox {
  if (rotation % 2 === 0) {
    // item is right side up or upside down
    return {
      left: position.x,
      top: position.y,
      bottom: height + position.y,
      right: width + position.x
    }
  }

  // item is rotated at a 90 degree angle (CW or CCW)
  const midX = position.x + (width / 2)
  const midY = position.y + (height / 2)

  return {
    left: midX - (height / 2),
    top: midY - (width / 2),
    right: midX + (height / 2),
    bottom: midY + (width / 2)
  }
}

/**
 * Computes the group bounding box for the given list of bounding boxes.
 *
 * @param boundingBoxes - list of bounding boxes contained
 * @returns computed bounding box
 */
function getGroupBoundingBox (boundingBoxes: BoundingBox[]): BoundingBox {
  return boundingBoxes.reduce((rect, boundingBox) => ({
    left: Math.min(rect.left, boundingBox.left),
    top: Math.min(rect.top, boundingBox.top),
    right: Math.max(rect.right, boundingBox.right),
    bottom: Math.max(rect.bottom, boundingBox.bottom)
  }), {
    left: Infinity,
    top: Infinity,
    right: 0,
    bottom: 0
  })
}

/**
 * Returns the center point of the given bounding box.
 */
function getBoundingBoxMidpoint (boundingBox: BoundingBox): Point {
  const width = boundingBox.right - boundingBox.left
  const height = boundingBox.bottom - boundingBox.top

  return {
    x: width / 2,
    y: height / 2
  }
}

/**
 * Given parent and child bounding boxes, determines the screen point (top left) of the child centered in the parent.
 *
 * @param parentBox
 * @param childBox
 * @param factor - any factor to round by
 */
function getCenteredScreenPoint (parentBox: BoundingBox, childBox: BoundingBox, factor: number): Point {
  const parentMidpoint = getBoundingBoxMidpoint(parentBox)
  const childWidth = childBox.right - childBox.left
  const childHeight = childBox.bottom - childBox.top

  return {
    x: roundToNth(parentMidpoint.x - (childWidth / 2), factor),
    y: roundToNth(parentMidpoint.y - (childHeight / 2), factor)
  }
}

export default {
  isInNeighborhood,
  getRadialSnapOffset,
  getOuterSnapOffset,
  getGridSnapPosition,
  areLinesIntersectingRectangle,
  isLineIntersectingRectangle,
  scaleBoundingBox,
  isOneOrTwoDimensional,
  getLinearBoundaries,
  getPointBoundary,
  hasIntersection,
  getBoundingBox,
  getGroupBoundingBox,
  getBoundingBoxMidpoint,
  getCenteredScreenPoint,
  getSnapOffset
}
