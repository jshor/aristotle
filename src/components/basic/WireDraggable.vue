<template>
  <svg
    class="wire"
    :class="{
      'wire--selected': isSelected,
      'wire--preview': isPreview,
      'wire--flash': flash
    }"
    :width="geometry.width"
    :height="geometry.height"
    aria-roledescription="connection"
  >
    <title>{{ label }}</title>
    <path
      class="wire__outline"
      fill="none"
      :transform="`translate(${Math.abs(geometry.minX)}, ${Math.abs(geometry.minY)})`"
      :d="geometry.path"
    />
    <path
      class="wire__display"
      :class="{
        'wire__display--on': sourceValue === LogicValue.TRUE,
        'wire__display--off': sourceValue === LogicValue.FALSE
      }"
      fill="none"
      :transform="`translate(${Math.abs(geometry.minX)}, ${Math.abs(geometry.minY)})`"
      :d="geometry.path"
    />
    <path
      v-if="isAnimated"
      class="wire__animation"
      :class="{
        'wire__animation--on': sourceValue === LogicValue.TRUE,
        'wire__animation--off': sourceValue === LogicValue.FALSE
      }"
      fill="none"
      :transform="`translate(${Math.abs(geometry.minX)}, ${Math.abs(geometry.minY)})`"
      :d="geometry.path"
    />
    <path
      class="wire__clickable"
      :class="{
        'wire__clickable--mobile': isMobile
      }"
      data-test="wire-clickable"
      fill="none"
      ref="wireRef"
      :transform="`translate(${Math.abs(geometry.minX)}, ${Math.abs(geometry.minY)})`"
      :d="geometry.path"
      @mousedown.left="onLeftMouseDown"
      @mouseup="onMouseUp"
      @touchstart="onTouchStart"
      @touchmove="onTouchDrag"
      @touchend="onTouchEnd"
      @touchcancel="dragEnd"
    />
  </svg>
</template>

<script setup lang="ts">
import { computed, ref } from 'vue'
import { useDraggable } from '@/composables/useDraggable'
import LogicValue from '@/types/enums/LogicValue'
import WireGeometry from '@/types/types/WireGeometry'
import Point from '@/types/interfaces/Point'
import isMobile from '@/utils/isMobile'
import { TOUCH_SHORT_HOLD_TIMEOUT } from '@/constants'

const props = defineProps<{
    /** The wire Bezier curve geometry. */
  geometry: WireGeometry
    /** Whether or not to display marching ants to indicate electrical flow direction. */
  isAnimated: boolean
    /** The logical value of the source port. */
  sourceValue: LogicValue
  /** The user-friendly label. */
  label: string
  /** Whether or not this connection is a preview (i.e., opaque in appearance). */
  isPreview?: boolean
  /** Whether or not this wire should display as selected (i.e., with a select shadow). */
  isSelected: boolean
  /** Whether or not the item should show a flash once to the user. */
  flash: boolean

}>()

const emit = defineEmits<{
  (e: 'add', point: Point, offset: Point): void
  (e: 'move', point: Point, offset: Point): void
  (e: 'added'): void
  (e: 'select', deselectAll?: boolean): void
  (e: 'deselect'): void
}>()

const wireRef = ref<SVGPathElement>()
const isTouchDragging = ref(false)
const allowTouchDrag = computed(() => isTouchDragging.value)

/**
 * Touch termination event handler.
 */
function onTouched ($event: TouchEvent, held: boolean) {
  if (held) {
    isTouchDragging.value = true

    return emit('add', {
      x: $event.touches[0].clientX,
      y: $event.touches[0].clientY
    }, { x: 0, y: 0 })
  }

  props.isSelected
    ? emit('deselect')
    : emit('select', false)
}

/**
 * Drag start event handler.
 */
function onDragStart (position: Point, offset: Point) {
  if (!isTouchDragging.value) {
    emit('add', position, offset)
  }
}

/**
 * Drag handler.
 */
function onDrag (position: Point, offset: Point) {
  emit('move', position, offset)
}

/**
 * Terminates the dragging operation.
 */
function onDragEnd () {
  isTouchDragging.value = false
  emit('added')
}

/**
 * Handles the left mouse down event.
 * This will select the item if it is not already selected.
 * The draggable `mousedown` event will continue to be propagated.
 */
function onLeftMouseDown ($event: MouseEvent) {
  onMouseDown($event)

  if (!props.isSelected) {
    emit('select', !$event.ctrlKey)
  }
}

/**
 * Handles touch dragging.
 */
function onTouchDrag ($event: TouchEvent) {
  if (isTouchDragging.value) {
    $event.stopPropagation()
  }

  onTouchMove($event)
}


const {
  onMouseDown,
  onMouseUp,
  onTouchStart,
  onTouchMove,
  onTouchEnd,
  dragEnd,
} = useDraggable({
  allowTouchDrag,
  longTouchTimeout: TOUCH_SHORT_HOLD_TIMEOUT,
  onTouched,
  onDragStart,
  onDrag,
  onDragEnd
})
</script>

<style lang="scss">
.wire {
  @include circuit-element;

  pointer-events: none;
  touch-action: none;
  position: absolute;

  &--preview {
    opacity: 0.5;
  }

  &__clickable {
    animation: none;
    pointer-events: visibleStroke;
    stroke-width: 16;
    cursor: pointer;

    &--mobile {
      stroke-width: 24;
    }
  }

  &__outline {
    stroke-width: 8;
    stroke-linecap: square;
    stroke: var(--color-secondary);
  }

  &__animation {
    stroke-width: 3;
    stroke-dasharray: 5;
    animation: animate1 500s infinite linear;
    stroke: var(--color-bg-primary);
    opacity: 0;

    &--on {
      opacity: 1;
    }
  }

  &__display {
    stroke-width: 6;
    stroke: var(--color-hi-z);
    transition: 0.25s stroke; // TODO

    &--on {
      stroke: var(--color-on);
      stroke: -webkit-focus-ring-color !important;
    }

    &--off {
      stroke: var(--color-off);
      stroke: -webkit-focus-ring-color !important;
    }
  }
}

// TODO: move to keyframes scss file?
@keyframes animate1 {
  from {
    stroke-dashoffset: 0;
  }
  to {
    stroke-dashoffset: -10000;
  }
}

@keyframes animate2 {
  from {
    stroke-dashoffset: 0;
  }
  to {
    stroke-dashoffset: 0;
  }
}

@keyframes flash { // TODO: make this a mixin
  0% {
    filter: inherit;
  }
  50% {
    filter: drop-shadow(0px 0px 20px var(--color-primary));
  }
  100% {
    filter: inherit;
  }
}
</style>
