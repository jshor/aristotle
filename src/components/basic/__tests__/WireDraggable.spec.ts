import { mount, VueWrapper } from '@vue/test-utils'
import WireDraggable from '../WireDraggable.vue'
import { createControlPoint } from '@/store/document/actions/__tests__/__helpers__'
import boundaries from '@/store/document/geometry/boundaries'
import renderLayout from '@/store/document/geometry/wire'
import { TOUCH_SHORT_HOLD_TIMEOUT } from '@/constants'

describe('Draggable wire component', () => {
  let wrapper: VueWrapper<typeof WireDraggable>

  beforeEach(() => {
    wrapper = mount(WireDraggable, {
      props: {
        geometry: renderLayout(createControlPoint(), createControlPoint())
      }
    })
  })

  afterEach(() => {
    jest.resetAllMocks()
  })

  it('should match the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render an animated wire when desired', async () => {
    await wrapper.setProps({ isAnimated: true })

    expect(wrapper.find('.wire__animation').exists()).toBe(true)
  })

  describe('when using touch gestures', () => {
    const getTouchEvent = (clientX: number, clientY: number): Touch => ({ clientX, clientY }) as Touch

    it('should emit `add` to add a control point after being held down in the same place for more than the short touch-hold timeout', async () => {
      jest.useFakeTimers()

      await wrapper
        .find('[data-test="wire-clickable"]')
        .trigger('touchstart', {
          touches: [new TouchEvent('touchstart')]
        })

      jest.advanceTimersByTime(TOUCH_SHORT_HOLD_TIMEOUT + 1)

      expect(wrapper.emitted()).toHaveProperty('add')
    })

    describe('when movement is insubstantial', () => {
      const x = 10
      const y = 20

      beforeEach(async () => {
        await wrapper
          .find('[data-test="wire-clickable"]')
          .trigger('touchend', {
            touches: [getTouchEvent(x, y)]
          })
      })

      it('should not emit `dragEnd`when the touch finger is released', async () => {
        expect(wrapper.emitted()).not.toHaveProperty('dragEnd')
      })

      it('should emit `select`', async () => {
        expect(wrapper.emitted()).toHaveProperty('select')
      })

      it('should emit `deselect` when the item is selected', async () => {
        await wrapper.setProps({ isSelected: true })
        await wrapper
          .find('[data-test="wire-clickable"]')
          .trigger('touchend', {
            touches: [getTouchEvent(x, y)]
          })
        expect(wrapper.emitted()).toHaveProperty('deselect')
      })
    })

    describe('when tapped and dragged', () => {
      const clientX = 10
      const clientY = 20
      const $event = {
        touches: [getTouchEvent(clientX, clientY)],
        stopPropagation: jest.fn()
      } as unknown as TouchEvent

      describe('adding ports', () => {
        beforeEach(() => {
          wrapper.vm.onTouched($event, false)
        })

        it('should not emit `add`', () => {
          expect(wrapper.emitted()).not.toHaveProperty('add')
        })

        it('should emit `add` with the given position and offset', () => {
          const position = { x: 10, y: 20 }
          const offset = { x: 1, y: 2 }

          wrapper.vm.onDragStart(position, offset)

          expect(wrapper.emitted()).toHaveProperty('add')
          expect(wrapper.emitted('add')).toHaveLength(1)
          expect(wrapper.emitted('add')![0]).toEqual([position, offset])
        })

        it('should not stop event propagation', () => {
          wrapper.vm.onTouchDrag($event)

          expect($event.stopPropagation).not.toHaveBeenCalled()
        })
      })

      it('should emit `select` when the wire is not selected', async () => {
        await wrapper.setProps({ isSelected: false })
        wrapper.vm.onTouched($event, false)

        expect(wrapper.emitted()).toHaveProperty('select')
      })

      it('should emit `deselect` when the wire is selected', async () => {
        await wrapper.setProps({ isSelected: true })
        wrapper.vm.onTouched($event, false)

        expect(wrapper.emitted()).toHaveProperty('deselect')
      })
    })

    describe('when held down', () => {
      const clientX = 10
      const clientY = 20
      const $event = {
        touches: [getTouchEvent(clientX, clientY)],
        stopPropagation: jest.fn()
      } as unknown as TouchEvent

      beforeEach(() => {
        wrapper.vm.onTouched($event, true)
      })

      it('should emit `add` with the position held', () => {
        expect(wrapper.emitted()).toHaveProperty('add')
        expect(wrapper.emitted('add')).toHaveLength(1)
        expect(wrapper.emitted('add')![0]).toEqual([{
          x: clientX,
          y: clientY
        }, { x: 0, y: 0 }])
      })

      it('should not emit `add` again afterwards when dragging has started', () => {
        wrapper.vm.onDragStart({ x: 10, y: 20 }, { x: 0, y: 0 })

        expect(wrapper.emitted('add')).toHaveLength(1)
      })

      it('should stop event propagation while dragging', () => {
        wrapper.vm.onTouchDrag($event)

        expect($event.stopPropagation).toHaveBeenCalled()
      })
    })

    describe('when left mouse is down', () => {
      it('should emit `select` when the wire is not selected', async () => {
        await wrapper.setProps({ isSelected: false })
        await wrapper
          .find('[data-test="wire-clickable"]')
          .trigger('mousedown', { clientX: 10, clientY: 20 })

        expect(wrapper.emitted()).toHaveProperty('select')
      })

      it('should not emit `select` when the wire is selected', async () => {
        await wrapper.setProps({ isSelected: true })
        await wrapper
          .find('[data-test="wire-clickable"]')
          .trigger('mousedown', { clientX: 10, clientY: 20 })

        expect(wrapper.emitted()).not.toHaveProperty('select')
      })
    })
  })

  describe('on dragging', () => {
    it('should emit `move`', () => {
      wrapper.vm.onDrag({ x: 10, y: 20 })

      expect(wrapper.emitted()).toHaveProperty('move')
      expect(wrapper.emitted('move')).toHaveLength(1)
      expect(wrapper.emitted('move')![0]).toEqual([{ x: 10, y: 20 }])
    })
  })

  describe('when dragging is complete', () => {
    it('should emit `added`', () => {
      wrapper.vm.onDragEnd()

      expect(wrapper.emitted()).toHaveProperty('added')
    })
  })

  describe('when the component is destroyed', () => {
    it('should remove the mousemove and mouseup methods', async () => {
      const spy = jest.spyOn(window, 'removeEventListener')

      await wrapper.unmount()

      expect(spy).toHaveBeenCalled()
      expect(spy).toHaveBeenCalledWith('mousemove', expect.any(Function))
      expect(spy).toHaveBeenCalledWith('mouseup', expect.any(Function))
    })
  })
})
