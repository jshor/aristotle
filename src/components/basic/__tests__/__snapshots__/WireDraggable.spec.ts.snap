// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Draggable wire component should match the snapshot 1`] = `
"<svg class="wire" width="15" height="15" aria-roledescription="connection">
  <title></title>
  <path class="wire__outline" fill="none" transform="translate(7.5, 7.5)" d="M 0,0 L 0,0"></path>
  <path class="wire__display" fill="none" transform="translate(7.5, 7.5)" d="M 0,0 L 0,0"></path>
  <!--v-if-->
  <path class="wire__clickable" data-test="wire-clickable" fill="none" transform="translate(7.5, 7.5)" d="M 0,0 L 0,0"></path>
</svg>"
`;
